### ISSUE TITLE
Questions on constructing pairs with distributed batch size in SimCLR

### ISSUE THREAD

**Issue by DianCh** [facebookresearch/vissl#474]
Hi! From the original SimCLR paper the performance benefits from large batch size, i.e., a large number of negative pairs for the NT-Xent loss:
<img width="527" alt="Screen Shot 2021-11-21 at 12 52 27 PM" src="https://user-images.githubusercontent.com/32043460/142778590-48869f8f-f593-452d-8abf-1cae36fbd378.png">

however, when we run training with DDP, the batch is divided into K sub-batches and the loss is calculated with only the samples from the sub-batch on each rank (i.e., each rank sees only N/K). In this case, how do we reproduce the original setting where the contrastive distribution is over 4096 examples?

Thank you!

**Comment by iseessel** [comment]
Hi @DianCh we all_gather the embeddings from each distributed rank: https://github.com/facebookresearch/vissl/blob/main/vissl/losses/simclr_info_nce_loss.py#L139. 

Our [config](https://github.com/facebookresearch/vissl/blob/main/configs/config/pretrain/simclr/simclr_8node_resnet.yaml) has 64 batchsize per gpu, 8 gpus, and 8 nodes, which is a global batchsize of 4096.

**Comment by DianCh** [comment]
Thank you @iseessel ! It's great to know that your gather implementation doesn't cut off the gradient, which I think answers exactly the main issue.

**Comment by BoPang1996** [comment]
Hi, I do not understand why the GatherLayer function "[all_reduce](https://github.com/facebookresearch/vissl/blob/64d705522b7d214d0f6913eaf65a70dff8b48ec4/vissl/utils/distributed_utils.py#L32)" the gradients in the backward function. In my understanding, for all the ranks, the "*grad" are the same, which is already carried all the gradients from all the samples. Thus this all_reduce op will enlarge the gradient by "dist.word_size()" times.
